### PR TITLE
API: Warn about dups in names for read_csv

### PR DESCRIPTION
**Pull request by gfyoung**
Title is self-explanatory.

xref #17095.

### CODE REVIEW

File: pandas/io/parsers.py [hunk 1]
```diff
@@ -384,6 +384,38 @@ def _validate_integer(name, val, min_val=0):
...
+        An array containing a list of the names used for the output DataFrame.
+    """
+
+    if names is None:
+        return
+
+    counts = {}
+    warn_dups = False
+
+    for name in names:
```
Comment (jreback) [inline pandas/io/parsers.py]: just use set intersection here

File: pandas/io/parsers.py [hunk 2]
```diff
@@ -384,6 +384,38 @@ def _validate_integer(name, val, min_val=0):
...
+        if name in counts:
+            warn_dups = True
+            break
+
+        counts[name] = True
+
+    if warn_dups:
+        msg = ("Duplicate names specified. This "
```
Comment (jreback) [inline pandas/io/parsers.py]: so are we deprecating this? then this should be a FutureWarning.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/parsers.py (hunk 1 above)
Comment (gfyoung) [reply pandas/io/parsers.py]: How so?  This is a fail-early method, which is why I chose it.

File: pandas/io/parsers.py (hunk 2 above)
Comment (gfyoung) [reply pandas/io/parsers.py]: Fair enough.  Done.

File: pandas/io/parsers.py (hunk 1 above)
Comment (jreback) [reply pandas/io/parsers.py]: simply check for ``len(names) !+ len(set(names)).``  much more idiomatic

File: pandas/io/parsers.py [hunk 3]
```diff
@@ -406,6 +438,10 @@ def _read(filepath_or_buffer, kwds):
     chunksize = _validate_integer('chunksize', kwds.get('chunksize', None), 1)
     nrows = _validate_integer('nrows', kwds.get('nrows', None))
 
+    # Check for duplicates in names.
+    names = kwds.get("names", None)
+    _check_dup_names(names)
```
Comment (jreback) [inline pandas/io/parsers.py]: call this ``_validate_names`` and have it return names, so its a similar patter to the other validators

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/parsers.py (hunk 1 above)
Comment (gfyoung) [reply pandas/io/parsers.py]: Fair enough.  Done.

File: pandas/io/parsers.py (hunk 3 above)
Comment (gfyoung) [reply pandas/io/parsers.py]: Done.

**Comment by gfyoung** [comment]
@jreback : All comments addressed and tests are green.  PTAL

**Comment by gfyoung** [comment]
@jreback @jorisvandenbossche : ping

File: doc/source/whatsnew/v0.21.0.txt [hunk 4]
```diff
@@ -283,6 +283,7 @@ Other API Changes
 - The Categorical constructor no longer accepts a scalar for the ``categories`` keyword. (:issue:`16022`)
 - Accessing a non-existent attribute on a closed :class:`~pandas.HDFStore` will now
   raise an ``AttributeError`` rather than a ``ClosedFileError`` (:issue:`16301`)
+- :func:`read_csv` now issues a ``UserWarning`` if the ``names`` parameter contains duplicates (:issue:`17095`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: should be FutureWarning

**Review by jreback — APPROVED** [review]
typo in whatsnew, otherwise lgtm. make sure this is on the deprecation list as well

File: doc/source/whatsnew/v0.21.0.txt (hunk 4 above)
Comment (gfyoung) [reply doc/source/whatsnew/v0.21.0.txt]: Doh!  My bad for not catching that.  Fixed.

**Comment by gfyoung** [comment]
Fixed typo and added to deprecation list.  Will merge on green then unless told otherwise.

**Comment by jreback** [comment]
@gfyoung wait for @jorisvandenbossche comment (as not sure if he commented here). IIRC a comment he made that having duplicate names is *ok* .

**Comment by gfyoung** [comment]
Sure thing.  FWIW, @jorisvandenbossche agreed with your suggestion, see his comment <a href="https://github.com/pandas-dev/pandas/pull/17095#discussion_r130438096">here</a>

@jorisvandenbossche : Any comments on this PR?

**Comment by gfyoung** [comment]
@jorisvandenbossche : ping if there any additional comments

**Comment by gfyoung** [comment]
@jreback : It's been a week, and I haven't heard anything from @jorisvandenbossche .  Still wait, or can we merge this PR?

**Comment by TomAugspurger** [comment]
I think Joris is off on holiday. I believe he's back next week.

**Comment by gfyoung** [comment]
@TomAugspurger : Ah!  I had a feeling that that was the case (I remember seeing an email about that).  I'll wait then until he gets back.

**Comment by gfyoung** [comment]
@jorisvandenbossche : friendly ping

**Comment by gfyoung** [comment]
@jorisvandenbossche : friendly ping

**Comment by jorisvandenbossche** [comment]
Can you remember me the rationale for deprecating this? 
Is it because we cannot actually handle it well?

Because I actually previously had a usecase where this proved useful (I had a non-informative column every other column, gave it the same name in `names` and then dropped the single name. But this specific case can of course easily be solved differently, by giving names like 'dummy1', 'dummy2', .. and then removing all columns that start with 'dummy')

**Comment by gfyoung** [comment]
@jorisvandenbossche : Here is what you said back in July <a href="https://github.com/pandas-dev/pandas/pull/17095#discussion_r130437981">here</a>.

Essentially, we are deprecating this behavior because `names` is a user-specified parameter, and passing in duplicate names *deliberately* only encourages buggy behavior.

**Comment by gfyoung** [comment]
@jorisvandenbossche : Any further thoughts on this?

**Comment by gfyoung** [comment]
@jorisvandenbossche : Any further thoughts on this?

**Comment by jorisvandenbossche** [comment]
Sorry for the slow response. 

So maybe a more general question: is it our intention to once fix `mangle_dupe_cols=True` ? (as currently actually only the default False works). If we do, I see no good reason to disallow duplicates in passed `names` vs names in the csv file.

**Comment by gfyoung** [comment]
@jorisvandenbossche : I think you meant the other way around.  `mangle_dupe_cols=True` is fully supported at this point.  It is `mangle_dupe_cols=False` that we disabled.

The reason for us discouraging duplicates in `names` is because duplicates are generally more error-prone.  Contrary to duplicates in a file, `names` is a *deliberate* choice.

@jreback : Thoughts?

**Comment by jorisvandenbossche** [comment]
>  I think you meant the other way around

yes, of course ... :-)

> Contrary to duplicates in a file, names is a deliberate choice.

That's true. But there are many other ways to deliberately make a dataframe with duplicate columns which we don't disallow anyway.

To be clear, in general I am all for a restricted scope of capabilities/possibilties. But in this case, limiting the abilities of `name` does not actually reduce code complexity, but increases it. As the duplicates are already perfectly handled by the code, so we are introducing a special case. Therefore I was wondering whether this is actually needed (user will see that the names are mangled anyhow).

**Comment by gfyoung** [comment]
@jorisvandenbossche : True that we'll see them mangled anyhow, but why the need to add complexity to just handle them in the first place?  I added the handling for duplicate `names` in an earlier PR because it was a bug, not to enhance support.

If the user really wants to have duplicate names, they can set it themselves and reading in the file, but I don't know if we want to actively encourage setting duplicate names to a read-in `DataFrame`.

**Comment by jorisvandenbossche** [comment]
Ah, I assumed that the mangling of `names` or the names from the header of the file was done using the same code path? That's not the case ? 
(to state it another way: once we remove the deprecation in this PR, we can actually remove more code than is added in this PR?)

**Comment by gfyoung** [comment]
See #17095 : it's a not a ridiculous amount of new logic that I added, but new logic nonetheless :smile:  Also, see @jreback comment in that PR <a href="https://github.com/pandas-dev/pandas/pull/17095/files#r129816735">here</a>

**Comment by gfyoung** [comment]
@jorisvandenbossche : Any updates on this?

**Comment by jreback** [comment]
so the only reason we have ``mangle_dupe_columns`` is to support duplicates in the first place. I think I stated we should deprecate that argument entirely, then I would allow duplicates in ``names`` but show a ``UserWarning`` if ``names`` contains duplicates.

So pretty much allow what is happening today but with a ``UserWarning`` and reducing the path complexity a bit (removing ``mangle_dupe_columes``).

Its not an error to have duplicates in ``names`` but I guess can't disallow it entirely.

**Comment by gfyoung** [comment]
@jreback : I don't recall you saying this before.  In addition, I think there has been user interest and *not* mangling in cases when the CSV file itself contains dupe names.  That being said, if we think making the warning less harsh is a good idea, I can do that.

File: doc/source/whatsnew/v0.21.0.txt (hunk 4 above)
Comment (gfyoung) [reply doc/source/whatsnew/v0.21.0.txt]: Changing back to `UserWarning` in light of later discussion.

**Comment by gfyoung** [comment]
@jreback : I made it issue a `UserWarning` instead.  PTAL.

**Comment by jreback** [comment]
lgtm. you might want to note in the doc-string the same warning.

**Comment by gfyoung** [comment]
> lgtm. you might want to note in the doc-string the same warning.

Sounds good.  I'll quickly add that.

**Comment by gfyoung** [comment]
@jreback : All is green.  PTAL.

File: pandas/io/parsers.py [hunk 5]
```diff
@@ -385,6 +385,33 @@ def _validate_integer(name, val, min_val=0):
     return val
 
 
+def _validate_names(names):
+    """
+    Check if the `names` parameter contains duplicates.
+
+    Currently, this function issues a warning if that is the case. In the
+    future, we will raise an error.
```
Comment (jreback) [inline pandas/io/parsers.py]: doc string  needs updating

**Review by jreback — APPROVED** [review]
lgtm modulo small comment

File: pandas/io/parsers.py (hunk 5 above)
Comment (gfyoung) [reply pandas/io/parsers.py]: Good catch.  Fixed.

**Comment by gfyoung** [comment]
@jreback : All is green.  PTAL.

**Comment by jreback** [comment]
thanks @gfyoung I think fine for now, we can always revisit if needed.

**Pull request merged**